File: src/arrays_and_strings/java/RotateMatrix.java
package arrays_and_strings.java;

import java.util.Scanner;

/**
 *
 */
public class RotateMatrix {
  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    int n = in.nextInt();

    int[][] matrix = new int[n][n];

    int start = 0;
    for (int i = 0; i < n; i++) {
      for (int j = 0; j < n; j++) {
        matrix[i][j] = start++;
      }
    }
    printMatrix(matrix);
    System.out.println();

    rotate(matrix);
    printMatrix(matrix);
  }

  private static void rotate(int[][] matrix) {
    if (matrix.length == 0) throw new IllegalArgumentException("Empty matrix");
    if (!isSquareMatrix(matrix)) throw new IllegalArgumentException("Not square matrix");

    int n = matrix.length;
    for (int i = 0; i < n / 2; i++) {
      // rotate by circle
      for (int j = i; j < n - i - 1; j++) {
        int temp = matrix[i][j];

        // left to top
        matrix[i][j] = matrix[n - j - 1][i];
        // bottom to left
        matrix[n - j - 1][i] = matrix[n - i - 1][n - j - 1];
        // right to bottom
        matrix[n - i - 1][n - j - 1] = matrix[j][n - i - 1];
        // top to right
        matrix[j][n - i - 1] = temp;
      }
    }
  }

  private static boolean isSquareMatrix(int[][] matrix) {
    int n = matrix.length;

    for (int[] arr : matrix) {
      if (n != arr.length) return false;
    }

    return true;
  }

  private static void printMatrix(int[][] matrix) {
    for (int[] arr : matrix) {
      for (int i : arr) {
        System.out.printf("%5d", i);
      }
      System.out.println();
    }
  }
}

File: src/arrays_and_strings/kotlin/RotateMatrix.kt
package arrays_and_strings.kotlin

/**
 *
 */
fun main(args: Array<String>) {
  val n = readLine().toString().toInt()

  val matrix = Array(n, { IntArray(n) })

  var count = 0
  for (i in matrix.indices) {
    for (j in matrix[i].indices) {
      matrix[i][j] = count++
    }
  }

  printMatrix(matrix)
  print("\n")
  rotate(matrix)
  printMatrix(matrix)
}

fun rotate(matrix: Array<IntArray>) {
  if (matrix.isEmpty()) throw IllegalArgumentException("Empty matrix")
  if (!isSquare(matrix)) throw IllegalArgumentException("Not square matrix")

  val n = matrix.size
  for (i in 0 until n / 2) {
    for (j in i until n - i - 1) {
      val temp = matrix[i][j]

      matrix[i][j] = matrix[n - j - 1][i]
      matrix[n - j - 1][i] = matrix[n - i - 1][n - j - 1]
      matrix[n - i - 1][n - j - 1] = matrix[j][n - i - 1]
      matrix[j][n - i - 1] = temp
    }
  }
}

fun printMatrix(matrix: Array<IntArray>) {
  matrix.forEach {
    it.forEach { print(String.format("%5d", it)) }
    print("\n")
  }
}

fun isSquare(matrix: Array<IntArray>): Boolean {
  val len = matrix.size
  matrix.forEach { if (it.size != len) return false }

  return true
}